src/controllers/softtakeover.h

```diff
@@ -1,56 +1,97 @@
+        m_dThreshold = threshold;
+    }
 
-    struct TestAccess;
+    // TODO (XXX): find a better testing solution than this TestAccess
+    // front-door coupled to `mixxx::Time`.
+    struct TestAccess {
+        static constexpr auto getTimeThreshold() {
+            return kSubsequentValueOverrideTime;
+        }
+        template<class Rep = mixxx::Time::rep,
+                class Period = mixxx::Time::period>
+        static void advanceTimePastThreshold(
+                std::chrono::duration<Rep, Period> offset =
+                        std::chrono::nanoseconds(0)) {
+            mixxx::Time::addTestTime(getTimeThreshold() + offset);
+        }
+    };
 
   private:
+    using ClockT = mixxx::Time;
     // If a new value is received within this amount of time, jump to it
     // regardless. This allows quickly whipping controls to work while retaining
     // the benefits of soft-takeover for slower movements.  Setting this too
     // high will defeat the purpose of soft-takeover.
-    static const mixxx::Duration kSubsequentValueOverrideTime;
+    static constexpr ClockT::duration kSubsequentValueOverrideTime = std::chrono::milliseconds(50);
+    static constexpr ClockT::time_point kFirstValueTime = ClockT::time_point::min();
 
-    mixxx::Duration m_time;
-    double m_prevParameter;
-    double m_dThreshold;
-};
-
-struct SoftTakeover::TestAccess {
-    static mixxx::Duration getTimeThreshold() {
-        return kSubsequentValueOverrideTime;
-    }
+    ClockT::time_point m_time{kFirstValueTime};
+    double m_prevParameter{0};
+    double m_dThreshold{kDefaultTakeoverThreshold};
 };
 
 class SoftTakeoverCtrl {
   public:
-    SoftTakeoverCtrl();
-    ~SoftTakeoverCtrl();
+    SoftTakeoverCtrl() = default;
 
     // Enable soft-takeover for the given Control.
     // This does nothing on a control that already has soft-takeover enabled.
-    void enable(ControlObject* control);
+    void enable(gsl::not_null<ControlPotmeter*> control);
     // Disable soft-takeover for the given Control
-    void disable(ControlObject* control);
+    void disable(ControlObject* control) {
+        m_softTakeoverHash.erase(control);
+    }
     // Check to see if the new value for the Control should be ignored
-    bool ignore(ControlObject* control, double newMidiParameter);
+    bool ignore(ControlObject* control, double newParameter) {
+        auto coIt = m_softTakeoverHash.find(control);
+        if (coIt == m_softTakeoverHash.end()) {
+            return false;
+        }
+        VERIFY_OR_DEBUG_ASSERT(control) {
+            return false;
+        }
+        return coIt->second.ignore(*control, newParameter);
+    }
     // Ignore the next supplied parameter
-    void ignoreNext(ControlObject* control);
+    void ignoreNext(ControlObject* control) {
+        auto coIt = m_softTakeoverHash.find(control);
+        if (coIt == m_softTakeoverHash.end()) {
+            return;
+        }
+
+        coIt->second.ignoreNext();
```

Comment:
wbout this?
```suggestion
        auto coIt = m_softTakeoverHash.find(control);
        if (coIt == m_softTakeoverHash.end()) {
            return;
        }
        auto [key, softTakeoverInstance] = *coIt;
        softTakeoverInstance.ignoreNext();
```

Reply:
This looks like a copied value, is it? What is the downside of the wrapper that returns a pointer? Normally we avoid non const references.

Reply:
> This looks like a copied value, is it?

Whoops, yes. make it `auto&` then.

> Normally we avoid non const references.

Does not apply here IMO. Or can you link me to the rationale behind this again?

> What is the downside of the wrapper that returns a pointer?

you force an unnecessarily unsafe API via the pointer that is being returned. You also introduce an extra also unnecessary nullcheck...

I also disagree that the anonymous pair hinders legibility much because thats just an integral part of `std::unordered_map`... Yes its ugly in the standard that they can't return a struct with named values, but thats how it is in the standard and I think its reasonable to expect to get used to it...

Reply:
Here is our rule: 
https://github.com/mixxxdj/mixxx/wiki/Coding-Guidelines#non-const-references

The "Whoops" here confirms that the pointer syntax is a reasonable choice for cases like here. 
The idea with the wrapper function helps to de-duplicate the code. A pointer with a null check is fully save and when we inline the wrapper as well it will disappear in the machine code. 

However if you prefer to keep a single function this is would be OK for me.

This one fully matches our guideline:
```suggestion
        auto it = m_softTakeoverHash.find(control);
        if (it == m_softTakeoverHash.end()) {
            return;
        }
        SoftTakeover* pSt = &(it->second);
        pSt->second.ignoreNext();
```

This one is kind of OK after our discussion
```suggestion
        auto it = m_softTakeoverHash.find(control);
        if (it == m_softTakeoverHash.end()) {
            return;
        }
        auto& [key, refSoftTakeover] = *it;
        refSoftTakeover.ignoreNext(); 
```

Reply:
I prefer the latter one as I don't see the purpose of taking a pointer (`&(it->second)`) just to dereference it again immediately after (`pSt->second`).

Reply:
> Here is our rule:
> https://github.com/mixxxdj/mixxx/wiki/Coding-Guidelines#non-const-references
> 
> The "Whoops" here confirms that the pointer syntax is a reasonable choice for cases like here.

I still don't agree with it. The rule seems to contrived and arbitrary... I will open a discussion on zulip.